src/Cooler.sol

```diff
@@ -148,74 +146,74 @@ contract Cooler is Clone {
+            remainder = repayment_ - loan.interestDue;
+            interestPaid = loan.interestDue;
+            loan.interestDue = 0;
         } else {
-            repayTo = address(this);
-            loan.unclaimed += repaid_;
+            loan.interestDue -= repayment_;
+            interestPaid = repayment_;
+        }
+
+        // We pay back only if user has paid back principle. This can be 0.
+        uint256 decollateralized;
+        if (remainder > 0) {
+            decollateralized = (loan.collateral * remainder) / loan.principle;
+            
+            loan.principle -= remainder;
+            loan.collateral -= decollateralized;
         }
 
         // Save updated loan info in storage.
         loans[loanID_] = loan;
 
-        // Transfer repaid debt back to the lender and (de)collateral back to the owner.
-        debt().safeTransferFrom(msg.sender, repayTo, repaid_);
-        collateral().safeTransfer(owner(), decollateralized);
+        // Transfer repaid debt back to the lender and collateral back to the owner if applicable
+        debt().safeTransferFrom(msg.sender, loan.recipient, repayment_);
+        if (decollateralized > 0) collateral().safeTransfer(owner(), decollateralized);
 
         // Log the event.
-        factory().newEvent(loanID_, CoolerFactory.Events.RepayLoan, repaid_);
+        factory().newEvent(loanID_, CoolerFactory.Events.RepayLoan, repayment_);
 
         // If necessary, trigger lender callback.
-        if (loan.callback) CoolerCallback(loan.lender).onRepay(loanID_, repaid_);
+        if (loan.callback) CoolerCallback(loan.lender).onRepay(loanID_, remainder, interestPaid);
+
         return decollateralized;
     }
 
-    /// @notice Roll a loan over with new terms.
-    ///         provideNewTermsForRoll must have been called beforehand by the lender.
-    /// @param  loanID_ index of loan in loans[].
-    function rollLoan(uint256 loanID_) external {
+    // Allow lender to extend loan for borrower. Any payments are done by the caller.
+    function extendLoanTerms(uint256 loanID_, uint8 times_) external {
         Loan memory loan = loans[loanID_];
 
+        if (msg.sender != loan.lender) revert OnlyApproved();
         if (block.timestamp > loan.expiry) revert Default();
-        if (!loan.request.active) revert NotRollable();
 
-        // Check whether rolling the loan requires pledging more collateral or not (if there was a previous repayment).
-        uint256 newCollateral = newCollateralFor(loanID_);
-        uint256 newDebt = interestFor(loan.amount, loan.request.interest, loan.request.duration);
-
-        // Update memory accordingly.
-        loan.amount += newDebt;
-        loan.collateral += newCollateral;
-        loan.expiry += loan.request.duration;
-        loan.request.active = false;
+        // Update loan terms to reflect the extension.
+        loan.expiry += loan.request.duration * times_;
+        loan.interestDue = interestFor(
```

Comment:
So, as I understand it:

1. Any amount repaid is ignored  when the loan is extended
2. At any point, the lender can extend the loan arbitrarily
3. The borrower does not get a penny's worth of collateral back until they've paid back all the extra interest that the lender unilaterally demanded

I cannot sign off on this fix.

Reply:
we can probably discuss this better in discord, but:
 1. Fair point, instead of `loan.request.amount` it should be `loan.principal`.
 2. 3. The premise was that it should always be favorable for a borrower to get an extension. Since the loans are fixed-term, it doesn't make sense to repay in advance, as you always need to pay the same interest. Following this idea, it makes even less sense to partially repay.

Nevertheless, I think your concern is still valid, as ideally, we want 0 trust assumptions between borrower and lender. We could slightly change the extension logic so that the interest due is not updated (therefore only benefiting the borrower with an extension). With this change, we would only need to slightly update the CH logic.

Reply:
- my idea of the remediation: https://github.com/ohmzeus/Cooler/pull/63

Reply:
So now Cooler loans function as an indefinite-period call option on gOhm? Not so allowing of attacks, but seems not great, economically.

I think this needs more thinking through before the final fix review.

Reply:
well this is exactly what the community voted... extending the loan is the same as repaying and getting another loan, we are just forcing them to prepay the interest in advance.